crates/astria-gossipnet/src/network.rs

```diff
@@ -131,43 +136,119 @@ impl Network {
+                None
+            };
+
             let behaviour = GossipnetBehaviour {
+                identify: identify::Behaviour::new(identify::Config::new(
+                    GOSSIPNET_PROTOCOL_ID.into(),
+                    public_key,
+                )),
                 gossipsub,
-                #[cfg(feature = "mdns")]
-                mdns,
                 ping: ping::Behaviour::default(),
+                mdns: mdns.into(),
+                kademlia: kademlia.into(),
             };
             SwarmBuilder::with_tokio_executor(transport, behaviour, local_peer_id).build()
         };
 
-        let listen_addr = format!("/ip4/0.0.0.0/tcp/{}", port);
+        let listen_addr = format!("/ip4/0.0.0.0/tcp/{}", self.port);
         swarm.listen_on(listen_addr.parse()?)?;
 
-        if let Some(addrs) = bootnodes {
+        if let Some(addrs) = self.bootnodes {
             addrs.iter().try_for_each(|addr| -> Result<_> {
-                debug!("dialing {:?}", addr);
-                let remote: Multiaddr = addr.parse()?;
-                swarm.dial(remote)?;
-                debug!("dialed {addr}");
+                let mut maddr: Multiaddr = addr.parse()?;
+                swarm.dial(maddr.clone())?;
+
+                let Some(peer_id) = maddr.pop() else {
+                    return Err(eyre!("failed to parse peer id from addr: {}", addr));
+                };
+
+                match peer_id {
+                    Protocol::P2p(peer_id) => {
+                        let peer_id = match PeerId::from_multihash(peer_id) {
+                            Ok(peer_id) => peer_id,
+                            Err(e) => {
+                                return Err(eyre!("failed to parse peer id from addr: {:?}", e));
+                            }
+                        };
+
+                        if let Some(kademlia) = swarm.behaviour_mut().kademlia.as_mut() {
+                            kademlia.add_address(&peer_id, maddr);
+                        }
+                    }
+                    _ => {
+                        return Err(eyre!("failed to parse peer id from addr: {}", addr));
+                    }
+                }
+
                 Ok(())
             })?;
         }
 
-        let multiaddr = Multiaddr::from_str(&format!("{}/p2p/{}", listen_addr, local_peer_id))?;
         Ok(Network {
-            multiaddr,
+            multiaddrs: vec![],
+            local_peer_id,
             swarm,
             terminated: false,
         })
     }
+}
+
+impl Default for NetworkBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub struct Network {
+    pub(crate) multiaddrs: Vec<Multiaddr>,
+    pub(crate) local_peer_id: PeerId,
+    pub(crate) swarm: Swarm<GossipnetBehaviour>,
+    pub(crate) terminated: bool,
+}
+
+impl Network {
+    pub fn multiaddrs(&self) -> Vec<Multiaddr> {
+        self.multiaddrs.clone()
+    }
+
+    pub async fn bootstrap(&mut self) -> Result<()> {
```

Comment:
also here. What does this fn do?

Reply:
see here: https://docs.rs/libp2p-kad/latest/libp2p_kad/struct.Kademlia.html#method.bootstrap it populates the routing table of the node's DHT